### PR TITLE
WIP: Immutable GitHub handle

### PR DESCRIPTION
**Pull request by izuroxx**
#413 
I've added custom validation in users model - immutable_github_handle & added test to it. it should not let the gh handle to be changed on update. before you could freely change it. (commit 61e43fa)
immutable_gh_before_disabling_field

in the end i've disabled the filed (commit 78326fc)
immutable gh

Now, few test are failing: 
spec/controllers/users_controller_spec.rb
spec/models/creates_application_from_draft_spec.rb
spec/controllers/application_drafts_controller_spec.rb
spec/models/application_draft_spec.rb

@ramonh i've tried to fix this, tired to exclude github_handle f.ex. in application_draft factory, but I'm a bit lost in tracing the problem. Would you give my some kind of a hint, direction, feedback about the method, tests?

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -17,6 +17,25 @@
   it { expect(subject).to allow_value('https://example.com').for(:homepage) }
   it { expect(subject).to_not allow_value('example.com').for(:homepage) }
 
+  describe 'immutable github handle validation' do
+    context 'when creating a new user' do
+      let(:new_user) { build(:user, github_handle: 'github_handle') }
+
+      it 'allows creating a user with given github handle' do
+        expect(new_user.valid?).to eq(true)
```
Comment (ramonh) [inline spec/models/user_spec.rb]: Minor thing here you might like. You can instead write:
`expect(new_user).to be_valid`

**Comment by izuroxx** [comment]
@ramonh i hope it all makes some sense :) test pass this time. there's also a email validation to this issue, which i've already started, but wanted to focus on this github_handle first

File: spec/factories/application_drafts.rb [hunk 2]
```diff
@@ -14,7 +14,8 @@
 
       after(:create) do |draft|
         draft.students.each do |student|
-          student.update(attributes_for :student, :applicant)
+          student_update_attributes = (attributes_for :student, :applicant).except(:github_handle)
+          student.update(student_update_attributes)
```
Comment (ramonh) [inline spec/factories/application_drafts.rb]: Great catch!

I wonder if there would be a way to clean this up so as to not have the application draft know about `github_handle`s. What do you think?

Comment (izuroxx) [inline spec/factories/application_drafts.rb]: yes, that would be great :) challenge accepted :+1:

Comment (ramonh) [inline spec/factories/application_drafts.rb]: Awesome! :v:

**Comment by carpodaster** [comment]
The changes made to `User.non_updateable_params` in 06c22b8 were alright. I don't think it works without the class method returning an `Array` and `Hash#except` being called with the splat'ed list of immutable attributes.

I think the reason why the tests failed is because `spec/controllers/users_controller_spec.rb:6` incorrectly removes github_**id** where it should be github_**handle**. You apply the changes in 06c22b8 again (only without `delete_keys_for_update` since it is not needed) and modify the spec as stated above? Do the tests still break?

File: app/models/user.rb [hunk 3]
```diff
@@ -125,7 +126,11 @@ def with_all_associations_joined
     def with_interest(interest)
       where(":interest = ANY(interested_in)", interest: interest)
     end
-  end
+
+    def non_updateable_params
```
Comment (carpodaster) [inline app/models/user.rb]: Naming suggestion: `immutable_attributes`

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -49,7 +49,7 @@ def create
 
   def update
     respond_to do |format|
-      if @user.update_attributes(user_params)
+      if @user.update_attributes(user_params.except(User.non_updateable_params))
```
Comment (carpodaster) [inline app/controllers/users_controller.rb]: I'm thinking: isn't the validation introduced below enough? If a user changes their GH handle, the validation will trigger and the record won't save. It will also display a meaningful error message.

Comment (izuroxx) [inline app/controllers/users_controller.rb]: understood :) i've deleted the except method. the github_handle field in a form is disabled, i've added disabled: true in the edit template (commit 78326fc). But when not disabled it displays github_handle can't be changed when trying to change it. shall i leave it disabled?

**Comment by izuroxx** [comment]
c8965d3 - changed `update attributes` in update action in user controller `immutable_github_handle` to` immutable_params`, deleted changed update attributes in update action in user controller
0a6f1fd - added `User.immutable_params.to_s` in user_controller_spec

File: app/models/user.rb [hunk 5]
```diff
@@ -127,7 +127,7 @@ def with_interest(interest)
       where(":interest = ANY(interested_in)", interest: interest)
     end
 
-    def non_updateable_params
+    def immutable_params
       :github_handle
```
Comment (carpodaster) [inline app/models/user.rb]: This should be an Array

File: app/models/user.rb (hunk 3 above)
Comment (izuroxx) [inline app/models/user.rb]: the `immutable_params` where created at first for future-proofing if there might be some changes with github_handle or more params added and also in order to call `User.immutable_params` in application_draft factory (+ user controller in update action - deleted in last commits). 
Application draft factory needed `User.immutable_params` to return symbol, but i wanted to test how to pass more symbols, and that's why i made an array and a splat.
As for failing tests - everything passes since	200f387 commit, now it does too.
As for `spec/controllers/users_controller_spec.rb:6` - "github_id" and avatar_info were in the test before I started to fix it. I guess it passed because i've had filtered Update_attributes in update action. But since i've deleted it -  I added `User.immutable_params.to_s` in the controller test to exclude github_handle. Hope it all makes some sense, if not please let me know how bad i've messed up this time ;p I will try to understad and fix it again :)

File: spec/controllers/users_controller_spec.rb [hunk 6]
```diff
@@ -3,7 +3,7 @@
 describe UsersController do
   render_views
 
-  let(:valid_attributes) { build(:user).attributes.except('github_id', 'avatar_url') }
+  let(:valid_attributes) { build(:user).attributes.except('github_id', 'avatar_url', User.immutable_params.to_s) }
```
Comment (carpodaster) [inline spec/controllers/users_controller_spec.rb]: If the class new method returns an array, it needs to be called with a `*`: 

```ruby
let(:valid_attributes) { build(:user).attributes.except('github_id', 'avatar_url', *User.immutable_params.map(&:to_s)) }

# Or:

let(:immutable_attributes) { User.immutable_params.map(&:to_s) + %w(github_id avatar_url) }
let(:valid_attributes) { build(:user).attributes.except(*immutable_attributes) }
```

File: spec/factories/application_drafts.rb [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
       after(:create) do |draft|
         draft.students.each do |student|
-          student_update_attributes = (attributes_for :student, :applicant).except(User.non_updateable_params)
+          student_update_attributes = (attributes_for :student, :applicant).except(User.immutable_params)
```
Comment (carpodaster) [inline spec/factories/application_drafts.rb]: Needs a splat (`*`) - see above

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (izuroxx) [inline app/controllers/users_controller.rb]: @carpodaster 
i've added disabled: true to the form in edit user template, so user can't access it  - shall i leave it that way or let it display error when changed

File: app/models/user.rb [hunk 8]
```diff
@@ -125,7 +126,11 @@ def with_all_associations_joined
     def with_interest(interest)
       where(":interest = ANY(interested_in)", interest: interest)
     end
-  end
+
+    def immutable_params
```
Comment (carpodaster) [inline app/models/user.rb]: As suggested before: could you name this `immutable_attributes` please? Rationale: in the context of this AR model, attributes is more appropriate. "params" hints at a controller context.

Comment (carpodaster) [inline app/models/user.rb]: The `def  … ` isn't properly indented :)

File: app/controllers/users_controller.rb (hunk 4 above)
Comment (carpodaster) [inline app/controllers/users_controller.rb]: Leaving it disabled is perfect. It doesn't obsolete the validation, though: remember all form data can be forged so it's not a sufficient to rely on the `disabled` HTML attribute: it's very easy to remove the disabled flag with your DOM inspector / editor of choice. The validation will catch those forged request.

TL;DR: it's all good with the form 👍

File: app/models/user.rb (hunk 8 above)
Comment (izuroxx) [inline app/models/user.rb]: :sob: true, it was suppose to be attributes, sorry for that

indentation - `def immutable_attributes` is in class << self block

Comment (carpodaster) [inline app/models/user.rb]: About the indentation: ooops! My bad. Sorry 😞

**Comment by carpodaster** [comment]
Looks superb now! Let's merge this.

**Comment by ramonh** [comment]
YAY! Nice one @izuroxx!

**Comment by izuroxx** [comment]
Thank You! :bow:
